docs/02.Tutorials/2.3.Pipeline-Explained.md

```diff
@@ -0,0 +1,592 @@
+- filter: validator
+  jumpIf:
+    invalid: mockRequestForProxy2
+- filter: mockRequest
+- filter: proxy1
+- filter: END
+- filter: mockRequest
+  alias: mockRequestForProxy2
+- filter: proxy2
+
+filters:
+- name: validator
+  kind: Validator
+  headers:
+    Content-Type:
+      values:
+      - application/json
+- name: mockRequest
+  kind: RequestBuilder
+  template: |
+    method: GET
+    url: /hello
+- name: proxy1
+  kind: Proxy
+  pools:
+  - servers:
+    - url: http://127.0.0.1:9095
+- name: proxy2
+  kind: Proxy
+  pools:
+  - servers:
+    - url: http://127.0.0.2:9095
+' | egctl apply -f -
+```
+
+### Namespace
+
+* The pipeline can handle multiple requests/responses, requests/responses are grouped by namespaces, and each namespace can have at most one request and one response.
+* We can set the `namespace` attribute of filters to let the filter know which request/response it should use.
+* `RequestBuilder` and `ResponseBuilder` can access all requests and responses, they output the result request/response to the namespace they are assigned.
+* The default namespace is `DEFAULT`.
+
+``` bash
+echo '
+name: pipeline-api
+kind: Pipeline
+
+flow:
+- filter: copyRequest
+  namespace: demo1
+- filter: copyRequest
+  namespace: demo2
+- filter: copyRequest
+  namespace: demo3
+- filter: proxy-demo1
+  namespace: demo1
+- filter: proxy-demo2
+  namespace: demo2
+- filter: proxy-demo3
+  namespace: demo3
+- filter: buildResponse
+
+filters:
+- name: copyRequest
+  kind: RequestBuilder
+  sourceNamespace: DEFAULT
+- name: proxy-demo1
+  kind: Proxy
+  pools:
+  - servers:
+    - url: https://demo1
+- name: proxy-demo2
+  kind: Proxy
+  pools:
+  - servers:
+    - url: https://demo2
+- name: proxy-demo3
+  kind: Proxy
+  pools:
+  - servers:
+    - url: https://demo3
+- name: buildResponse
+  kind: ResponseBuilder
+  template: |
+    statusCode: 200
+    body: [{{.responses.demo1.Body}}, {{.response.demo2.Body}}, {{.response.demo3.Body}}]
+' | egctl create -f -
+```
+
+## Usage
+
+### GlobalFilter
+GlobalFilter is a special pipeline that can be executed before or/and after all pipelines in a server. For example:
+
+```yaml
+name: globalFilter-example
+kind: GlobalFilter
+beforePipeline:
+  flow:
+  - filter: validator
+
+  filters:
+  - name: validator
+    kind: Validator
+    ...
+
+afterPipeline:
+  ...
+
+---
+name: server-example
+kind: HTTPServer
+globalFilter: globalFilter-example
+...
+```
+In this case, all requests in `HTTPServer` `server-example` go through `GlobalFilter` `globalFilter-example` before and after executing any other pipelines.
+
+### Load Balancer
+
+The reverse proxy is the common middleware that is accessed by clients, forwards them to backend servers. The reverse proxy is a very core role played by Easegress.
+
+The filter `Proxy` is the filter to fire requests to backend servers. It contains servers group under load balance, whose policy support `roundRobin`, `random`, `weightedRandom`, `ipHash`, `headerHash`.
```

Comment:
It is better to refer to the balance reference to check all specs of all policies.

Reply:
Updated.